cmd/ingester/app/consumer/consumer.go

```diff
@@ -42,6 +43,9 @@ type Consumer struct {
 	internalConsumer consumer.Consumer
 	processorFactory ProcessorFactory
 
+	deadlockDetectorFactory      deadlockDetectorFactory
```

Comment:
I think it would be cleaner and easier to understand if you had a top-level `deadlockDetector`, which can create `partitionDeadlockDetector` as needed (implementation-wide the former may contain the latter for pId=-1). So factory is only used once to create the top-level detector, and the factory does not need to be stored in the Consumer.

Reply:
Doesn't this mean that the top level `deadlockDetector` also has the responsibilities of the Factor? (That being said, I think it might be cleaner design)

Reply:
👍